ocrd_tesserocr/fontshape.py

```diff
@@ -0,0 +1,189 @@
+
+    def _process_lines(self, tessapi, textlines, region_image, region_coords):
+        LOG = getLogger('processor.TesserocrFontShape')
+        for line in textlines:
+            line_image, line_coords = self.workspace.image_from_segment(
+                line, region_image, region_coords)
+            LOG.debug("Recognizing text in line '%s'", line.id)
+            words = line.get_Word()
+            if not words:
+                LOG.warning("Line '%s' contains no words", line.id)
+            else:
+                self._process_words(tessapi, words, line_image, line_coords)
+
+    def _process_words(self, tessapi, words, line_image, line_coords):
+        LOG = getLogger('processor.TesserocrFontShape')
+        for word in words:
+            word_image, word_coords = self.workspace.image_from_segment(
+                word, line_image, line_coords)
+            if self.parameter['padding']:
+                tessapi.SetImage(pad_image(word_image, self.parameter['padding']))
+            else:
+                tessapi.SetImage(word_image)
+            tessapi.SetPageSegMode(PSM.SINGLE_WORD)
+            #tessapi.SetPageSegMode(PSM.RAW_LINE)
+            tessapi.Recognize()
+            result_it = tessapi.GetIterator()
+            if not result_it or result_it.Empty(RIL.WORD):
+                LOG.warning("No text in word '%s'", word.id)
+                continue
+            LOG.debug("Decoding text in word '%s'", word.id)
+            # trigger recognition
+            word_text = result_it.GetUTF8Text(RIL.WORD)
+            LOG.debug('Word "%s" detected "%s"', word.id, word_text)
+            textequiv = word.get_TextEquiv()
+            if textequiv:
+                LOG.info('Word "%s" annotated "%s" / detected "%s"',
+                         word.id, textequiv[0].Unicode, word_text)
+            word_attributes = result_it.WordFontAttributes()
+            if word_attributes:
+                #LOG.debug("found font attributes: {}".format(word_attributes))
+                word_style = TextStyleType(
+                    fontSize=word_attributes['pointsize']
+                    if 'pointsize' in word_attributes else None,
+                    fontFamily=word_attributes['font_name']
+                    if 'font_name' in word_attributes else None,
+                    bold=word_attributes['bold']
+                    if 'bold' in word_attributes else None,
+                    italic=word_attributes['italic']
+                    if 'italic' in word_attributes else None,
+                    underlined=word_attributes['underlined']
+                    if 'underlined' in word_attributes else None,
+                    monospace=word_attributes['monospace']
+                    if 'monospace' in word_attributes else None,
+                    serif=word_attributes['serif']
+                    if 'serif' in word_attributes else None)
+                word.set_TextStyle(word_style) # (or somewhere in custom attribute?)
```

Comment:
Excellent.

I think `<TextStyle>` is the right place, c.f. https://ocr-d.de/en/spec/page#textstyle

Reply:
Yes, but maybe we should also "propagate" it to the parents' `TextStyle` analogously to `TextEquiv`.

The problem with word-wise recognition is that it's not as reliable. But if we allow the pre-LSTM models to operate on the complete line, they will (in general) come up with a different segmentation than the one annotated in PAGE from some LSTM recognizer.

One improvement I had in mind was cropping word images from line y and word x, so that words without ascenders or descenders (like `von`) are not considered larger than others merely because they are cropped tighter.

Reply:
Tried this, and gave up: It does not help getting more homogeneous font sizes, but increases the likelihood of having intruding components from neighouring lines in the word before recognition (which also impairs font attribute quality).